FirebaseSessions/Sources/Settings.swift

```diff
@@ -0,0 +1,173 @@
+    } catch {
+      Logger.logError("[Settings] Error: \(error)")
+    }
+    return nil
+  }
+}
+
+extension NSLocking {
+  func synchronized<T>(_ closure: () throws -> T) rethrows -> T {
+    lock()
+    defer { unlock() }
+    return try closure()
+  }
+}
+
+protocol SettingsProtocol {
+  func loadCache(googleAppID: String, currentTime: Date)
+  var sessionsEnabled: Bool { get }
+  var samplingRate: Double { get }
+  var sessionTimeout: TimeInterval { get }
+}
+
+class Settings: SettingsProtocol {
+  private static let cacheDurationSecondsDefault: TimeInterval = 60 * 60
+  private let fileManager: SettingsFileManagerProtocol
+  private let appInfo: ApplicationInfoProtocol
+  // Underscored variables may be accessed in different threads,
+  private let lock = NSLock()
+  private var _settingsDictionary: [String: AnyObject]?
+  private var _isCacheKeyExpired: Bool
+  struct CacheKey: Codable {
+    var createdAt: Date
+    var googleAppID: String
+    var appVersion: String
+  }
+
+  var settingsDictionary: [String: AnyObject]? {
+    lock.synchronized {
+      _settingsDictionary
+    }
+  }
+
+  var sessionsEnabled: Bool {
+    guard let enabled = settingsDictionary?["sessions_enabled"]?.boolValue else {
+      return true
+    }
+    return enabled
+  }
+
+  var samplingRate: Double {
+    guard let rate = settingsDictionary?["sampling_rate"]?.doubleValue else {
+      return 1
+    }
+    return rate
+  }
+
+  var sessionTimeout: TimeInterval {
+    guard let timeout = settingsDictionary?["session_timeout"]?.doubleValue else {
+      return 30 * 60
+    }
+    return timeout
+  }
+
+  var isCacheExpired: Bool {
+    guard settingsDictionary != nil else {
+      return true
+    }
+    return lock.synchronized {
+      return _isCacheKeyExpired
+    }
+  }
+
+  private var cacheDurationSeconds: TimeInterval {
+    guard let duration = settingsDictionary?["cache_duration"]?.doubleValue else {
+      return Settings.cacheDurationSecondsDefault
+    }
+    return duration
+  }
+
+  init(fileManager: SettingsFileManagerProtocol = SettingsFileManager(),
+       appInfo: ApplicationInfoProtocol) {
+    self.fileManager = fileManager
+    _isCacheKeyExpired = false
+    self.appInfo = appInfo
+  }
+
+  func loadCache(googleAppID: String, currentTime: Date) {
+    guard let cacheData = fileManager.data(contentsOf: fileManager.settingsCacheContentPath) else {
+      Logger.logDebug("[Settings] No settings were cached")
+      return
+    }
+    guard let parsedDictionary = cacheData.dictionaryValue else {
+      removeCache()
+      return
+    }
+    lock.synchronized {
+      _settingsDictionary = parsedDictionary
+    }
+    guard let cacheKeyData = fileManager.data(contentsOf: fileManager.settingsCacheKeyPath),
+          let cacheKey = cacheKeyData.cacheKeyValue else {
+      Logger.logError("[Settings] Could not load settings cache key")
+      removeCache()
+      return
+    }
+    guard cacheKey.googleAppID == googleAppID else {
```

Comment:
Is this for the support for multi-apps? Does Crashlytics support multi-apps situation?

Reply:
It's for developers registering a new app id, not for multi-apps.